registries/common.go

```diff
@@ -53,21 +81,106 @@ func (u *URL) FromString(raw string) error {
 	}
 	u.protocol = rawURL.Scheme
 	u.host = rawURL.Host
+	u.interfaceName = strings.TrimPrefix(rawURL.Path, "/")
 	u.params = rawURL.Query()
 	return nil
 }
 
+func (u *URL) ToString() string {
+	paramsPart, _ := url.QueryUnescape(u.params.Encode())
+	raw := fmt.Sprintf("%s://%s/%s?%s", u.protocol, u.host, u.interfaceName, paramsPart)
+	return url.QueryEscape(raw)
+}
+
 func (u *URL) ToInstance() discovery.Instance {
 	weight := DefaultDubboServiceWeight
-	if weightStr := u.params.Get("weight"); weightStr != "" {
+	if weightStr := u.params.Get(dubboInternalWeightKey); weightStr != "" {
 		if weightParam, err := strconv.Atoi(weightStr); err == nil {
 			weight = weightParam
 		}
 	}
 	params := map[string]string{
-		DubboServiceProtocolKey: u.protocol,
-		DubboServiceGroupKey:    u.params.Get("group"),
-		DubboServiceVersionKey:  u.params.Get("version"),
+		DubboServiceGroupKey:   u.params.Get(dubboInternalGroupKey),
+		DubboServiceVersionKey: u.params.Get(dubboInternalVersionKey),
 	}
 	return discovery.NewInstance("tcp", u.host, weight, params)
 }
+
+func (u *URL) FromInfo(info *registry.Info) error {
+	u.protocol = DefaultProtocol
+	if err := u.checkAndSetHost(info.Addr.String()); err != nil {
+		return err
+	}
+	if err := u.filterAndSetParams(info.Tags); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u *URL) GetRegistryServiceKey(registryGroup string) string {
+	return fmt.Sprintf(RegistryServicesKeyTemplate, registryGroup, u.interfaceName)
+}
+
+func (u *URL) checkAndSetHost(addr string) error {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return fmt.Errorf("parse addr %s error: %s", addr, err)
+	}
+	if port == "" {
+		return fmt.Errorf("addr %s missing port", addr)
+	}
+	if host == "" || host == "::" {
+		ipv4, err := getLocalIPV4Address()
+		if err != nil {
+			return fmt.Errorf("get local ipv4 error, cause %s", err)
+		}
+		u.host = ipv4 + ":" + port
+	} else {
+		u.host = host + ":" + port
```

Comment:
Seems that IPv6 address is not dealt with properly.

```
SplitHostPort splits a network address of the form "host:port", "host%zone:port", "[host]:port" or "[host%zone]:port" into host or host%zone and port.

A literal IPv6 address in hostport must be enclosed in square brackets, as in "[::1]:80", "[::1%lo0]:80".
```
https://pkg.go.dev/net#SplitHostPort

Reply:
Ack.